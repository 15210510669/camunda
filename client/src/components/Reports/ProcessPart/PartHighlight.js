import React from 'react';

import './PartHighlight.css';

export default class PartHighlight extends React.Component {
  componentDidMount() {
    // add highlighted sequence flow end marker
    const {viewer} = this.props;

    this.defs = viewer._container.querySelector('defs');
    this.highlightMarker = this.defs.querySelector('marker').cloneNode(true);

    this.highlightMarker.setAttribute('id', 'sequenceflow-end-highlight');

    this.defs.appendChild(this.highlightMarker);
  }

  render() {
    const {nodes, viewer} = this.props;
    const reachableNodes = new Set();

    if (nodes.length === 2 && typeof nodes[0] === 'object') {
      // only if start and end are selected and are resolved to bpmn-js objects

      const graph = {};
      const unprocessed = [nodes[0]];

      // phase 1: construct reachability graph from start node
      while (unprocessed.length > 0) {
        const current = unprocessed.shift();

        current.outgoing &&
          current.outgoing.forEach(connection => {
            const targetId = connection.targetRef.id;

            if (!graph[targetId]) {
              graph[targetId] = [];
              unprocessed.push(connection.targetRef);
            }

            graph[targetId].push({node: current.id, via: connection.id});
          });
      }

      // phase 2: use backlinks from end node to find nodes between start and end
      if (graph[nodes[1].id]) {
        const unprocessed = [nodes[1].id];

        while (unprocessed.length > 0) {
          const current = unprocessed.shift();
          reachableNodes.add(current);

          graph[current] &&
            graph[current].forEach(({node, via}) => {
              reachableNodes.add(via);
              if (!reachableNodes.has(node) && !unprocessed.includes(node)) {
                unprocessed.push(node);
              }
            });
        }
      }
    }

    const canvas = viewer.get('canvas');

    // adjust markers
    viewer.get('elementRegistry').forEach(({businessObject, id}) => {
      canvas.removeMarker(id, 'PartHighlight');
      if (reachableNodes.has(businessObject.id)) {
        canvas.addMarker(id, 'PartHighlight');
      }
    });

    return null;
  }

  componentWillUnmount() {
    const {viewer} = this.props;

    viewer.get('elementRegistry').forEach(({id}) => {
      viewer.get('canvas').removeMarker(id, 'PartHighlight');
    });

    this.defs.removeChild(this.highlightMarker);
  }
}
