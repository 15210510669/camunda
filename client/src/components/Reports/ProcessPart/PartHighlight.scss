@import 'shared-styles.scss';

.PartHighlight {
  @include diagram-highlight;
}
