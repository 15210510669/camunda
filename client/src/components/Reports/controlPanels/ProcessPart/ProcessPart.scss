.ProcessPartModal .Modal__content {
  display: flex;
  flex-direction: column;

  .diagram-container {
    display: flex;
    flex-grow: 1;
  }

  .Message {
    margin-top: 8px;
  }
}

.ProcessPart__current {
  cursor: pointer;
  white-space: pre-wrap;
}
