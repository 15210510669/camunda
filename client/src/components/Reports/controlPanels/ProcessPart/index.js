export {default as ProcessPart} from './ProcessPart';
