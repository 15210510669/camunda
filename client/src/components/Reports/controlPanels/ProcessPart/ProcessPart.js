/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React from 'react';
import {Button, Modal, BPMNDiagram, ClickBehavior, ActionItem, Message} from 'components';

import PartHighlight from './PartHighlight';

import './ProcessPart.scss';

export default class ProcessPart extends React.Component {
  state = {
    modalOpen: false,
    start: null,
    end: null,
    hasPath: true
  };

  render() {
    return (
      <React.Fragment>
        {this.renderButton()}
        {this.renderPart()}
        {this.renderModal()}
      </React.Fragment>
    );
  }

  renderButton() {
    if (!this.props.processPart) {
      return <Button onClick={this.openModal}>Process Instance Part</Button>;
    }
  }

  renderFlowNodeName = id => {
    return this.props.flowNodeNames ? this.props.flowNodeNames[id] || id : id;
  };

  renderPart() {
    if (this.props.processPart) {
      return (
        <div onClick={this.openModal} className="ProcessPart__current">
          <ActionItem
            onClick={evt => {
              evt.stopPropagation();
              this.props.update(null);
            }}
          >
            Only regard part between{' '}
            <span className="FilterList__value">
              {this.renderFlowNodeName(this.props.processPart.start)}
            </span>
            <span> and </span>
            <span className="FilterList__value">
              {this.renderFlowNodeName(this.props.processPart.end)}
            </span>
          </ActionItem>
        </div>
      );
    }
  }

  setHasPath = hasPath => {
    if (this.state.hasPath !== hasPath) {
      // this is called during render of PartHighlight. We cannot update state during a render, so we do it later
      window.setTimeout(() => this.setState({hasPath}));
    }
  };

  renderModal() {
    const {start, end, hasPath, modalOpen} = this.state;

    const selection = [start, end].filter(v => v);
    return (
      <Modal
        open={modalOpen}
        onClose={this.closeModal}
        onConfirm={this.isValid() ? this.applyPart : undefined}
        size="max"
        className="ProcessPartModal"
      >
        <Modal.Header>Set Process Instance Part</Modal.Header>
        <Modal.Content>
          <span>
            Only regard the process instance part between{' '}
            <ActionItem
              disabled={!start}
              onClick={() => this.setState({start: null, hasPath: true})}
            >
              {start ? start.name || start.id : 'Please select start'}
            </ActionItem>{' '}
            and{' '}
            <ActionItem disabled={!end} onClick={() => this.setState({end: null, hasPath: true})}>
              {end ? end.name || end.id : 'Please select end'}
            </ActionItem>
          </span>
          {start && end && !hasPath && (
            <Message type="warning">
              You selected two nodes, but there is no obvious connection between those nodes. Report
              results may be empty or misleading.
            </Message>
          )}
          <div className="diagram-container">
            <BPMNDiagram xml={this.props.xml}>
              <ClickBehavior
                setSelectedNodes={this.setSelectedNodes}
                onClick={this.selectNode}
                selectedNodes={selection}
              />
              <PartHighlight nodes={selection} setHasPath={this.setHasPath} />
            </BPMNDiagram>
          </div>
        </Modal.Content>
        <Modal.Actions>
          <Button onClick={this.closeModal}>Cancel</Button>
          <Button type="primary" color="blue" onClick={this.applyPart} disabled={!this.isValid()}>
            Apply
          </Button>
        </Modal.Actions>
      </Modal>
    );
  }

  openModal = () => this.setState({modalOpen: true, ...this.props.processPart});
  closeModal = () => this.setState({modalOpen: false, start: null, end: null, hasPath: true});

  setSelectedNodes = ([start, end]) => {
    this.setState({start, end});
  };

  selectNode = node => {
    if (this.state.start === node) {
      return this.setState({start: null, hasPath: true});
    }
    if (this.state.end === node) {
      return this.setState({end: null, hasPath: true});
    }
    if (!this.state.start) {
      return this.setState({start: node});
    } else {
      return this.setState({end: node});
    }
  };

  applyPart = () => {
    this.props.update({start: this.state.start.id, end: this.state.end.id});
    this.closeModal();
  };

  isValid = () => this.state.start && this.state.end;
}
