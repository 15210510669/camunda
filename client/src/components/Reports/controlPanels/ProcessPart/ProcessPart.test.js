import React from 'react';

import ProcessPart from './ProcessPart';

import {mount} from 'enzyme';

jest.mock('components', () => {
  const Modal = props => <div id="modal">{props.open && props.children}</div>;
  Modal.Header = props => <div id="modal_header">{props.children}</div>;
  Modal.Content = props => <div id="modal_content">{props.children}</div>;
  Modal.Actions = props => <div id="modal_actions">{props.children}</div>;

  return {
    Modal,
    Button: props => <button {...props} />,
    BPMNDiagram: () => <div>BPMNDiagram</div>,
    ClickBehavior: () => <div>ClickBehavior</div>,
    ActionItem: ({onClick, children}) => (
      <div>
        <button className="clearBtn" onClick={onClick}>
          X
        </button>
        {children}
      </div>
    ),
    Message: ({children}) => <div className="message">{children}</div>
  };
});

jest.mock('./PartHighlight', () => () => null);

it('should display a button if no process part is set', () => {
  const node = mount(<ProcessPart />);

  expect(
    node.findWhere(n => n.type() === 'button' && n.text() === 'Process Instance Part')
  ).toBePresent();
});

it('should not display the button is process part is set', () => {
  const node = mount(<ProcessPart processPart={{start: 'a', end: 'b'}} />);

  expect(
    node.findWhere(n => n.type() === 'button' && n.text() === 'Process Instance Part')
  ).not.toBePresent();
});

it('should show a preview of the process part', () => {
  const node = mount(
    <ProcessPart
      processPart={{start: 'a', end: 'b'}}
      flowNodeNames={{a: 'Start Node', b: 'End Node'}}
    />
  );

  expect(node).toIncludeText('Only regard part between Start Node and End Node');
});

it('should remove the process part', () => {
  const spy = jest.fn();
  const node = mount(<ProcessPart processPart={{start: 'a', end: 'b'}} update={spy} />);

  node.find('button').simulate('click');

  expect(spy).toHaveBeenCalledWith(null);
});

it('should open a modal when clicking the button', () => {
  const node = mount(<ProcessPart processPart={{start: 'a', end: 'b'}} />);

  node.find('.ProcessPart__current').simulate('click');

  expect(node.state('modalOpen')).toBe(true);
});

it('should show the bpmn diagram', () => {
  const node = mount(<ProcessPart processPart={{start: 'a', end: 'b'}} />);

  node.find('.ProcessPart__current').simulate('click');

  expect(node.find('BPMNDiagram')).toBePresent();
});

it('should show the id of the selected node if it does not have a name', () => {
  const node = mount(<ProcessPart />);

  node.setState({
    modalOpen: true,
    start: {id: 'startId', name: 'Start Name'},
    end: {id: 'endId'}
  });

  expect(node.find('#modal_content')).toIncludeText('Start Name');
  expect(node.find('#modal_content')).toIncludeText('endId');
});

it('should deselect a node when it is clicked and already selected', () => {
  const node = mount(<ProcessPart />);

  const flowNode = {id: 'nodeId', name: 'Some node'};

  node.setState({
    modalOpen: true,
    start: flowNode
  });

  node.instance().selectNode(flowNode);

  expect(node.state('start')).toBe(null);
});

it('should display a warning if there is no path between start and end node', () => {
  const node = mount(<ProcessPart />);

  const flowNode = {id: 'nodeId', name: 'Some node'};

  node.setState({
    modalOpen: true,
    start: flowNode,
    end: {id: 'anId'},
    hasPath: false
  });

  expect(node.find('.message')).toIncludeText('Report results may be empty or misleading.');
});
