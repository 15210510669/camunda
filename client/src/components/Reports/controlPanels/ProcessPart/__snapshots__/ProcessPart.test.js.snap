// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should show a preview of the process part 1`] = `
<ActionItem
  onClick={[Function]}
  onEdit={[Function]}
>
  <Tag
    size="sm"
    type="blue"
  >
    Process Instance Part
  </Tag>
  <div>
    <b>
      Start Node
    </b>
    <span>
       
      and
       
    </span>
    <b>
      b
    </b>
  </div>
</ActionItem>
`;
